File: src/skills.js
const fs = require('fs');
const path = require('path');
const os = require('os');
const YAML = require('yaml');

const MAX_SKILL_FILE_BYTES = 1024 * 1024;

function expandHome(inputPath) {
  const value = String(inputPath || '').trim();
  if (!value || value === '~') return os.homedir();
  if (value.startsWith('~/')) return path.join(os.homedir(), value.slice(2));
  return value;
}

function isWithin(parentPath, childPath) {
  const relative = path.relative(parentPath, childPath);
  return relative === '' || (!relative.startsWith('..') && !path.isAbsolute(relative));
}

function parseFrontmatter(content, fallbackName) {
  const normalized = String(content || '').replace(/\r\n/g, '\n');
  const match = normalized.match(/^---\s*\n([\s\S]*?)\n---\s*(?:\n|$)/);
  let metadata = {};

  if (match) {
    try {
      metadata = YAML.parse(match[1]) || {};
    } catch {
      metadata = {};
    }
  }

  const name = typeof metadata.name === 'string' && metadata.name.trim()
    ? metadata.name.trim()
    : fallbackName;
  const description = typeof metadata.description === 'string'
    ? metadata.description.replace(/\s+/g, ' ').trim()
    : '';

  return { name, description };
}

function directPackageRoots(nodeModulesPath) {
  if (!fs.existsSync(nodeModulesPath)) return [];
  const roots = [];

  for (const entry of fs.readdirSync(nodeModulesPath, { withFileTypes: true })) {
    if (!entry.isDirectory() || entry.name.startsWith('.')) continue;
    const entryPath = path.join(nodeModulesPath, entry.name);

    if (entry.name.startsWith('@')) {
      for (const scoped of fs.readdirSync(entryPath, { withFileTypes: true })) {
        if (scoped.isDirectory()) roots.push(path.join(entryPath, scoped.name));
      }
    } else {
      roots.push(entryPath);
    }
  }

  return roots;
}

class SkillRegistry {
  constructor(getFsRoot = () => '~/') {
    this.getFsRoot = getFsRoot;
    this.skills = new Map();
  }

  candidateRoots() {
    const candidates = [];
    const add = (skillsRoot, source) => {
      if (!skillsRoot) return;
      const absolute = path.resolve(expandHome(skillsRoot));
      if (fs.existsSync(absolute) && fs.statSync(absolute).isDirectory()) {
        candidates.push({ root: absolute, source });
      }
    };

    const fsRoot = path.resolve(expandHome(this.getFsRoot()));
    add(path.join(fsRoot, '.agents', 'skills'), 'workspace');
    add(path.join(os.homedir(), '.agents', 'skills'), 'user');
    add(path.join(os.homedir(), '.pi', 'agent', 'skills'), 'pi');

    const extensionsRoot = path.join(os.homedir(), '.pi', 'agent', 'extensions');
    if (fs.existsSync(extensionsRoot)) {
      for (const entry of fs.readdirSync(extensionsRoot, { withFileTypes: true })) {
        if (entry.isDirectory()) {
          add(path.join(extensionsRoot, entry.name, 'skills'), `pi-extension:${entry.name}`);
        }
      }
    }

    const piNodeModules = path.join(os.homedir(), '.pi', 'agent', 'npm', 'node_modules');
    for (const packageRoot of directPackageRoots(piNodeModules)) {
      add(path.join(packageRoot, 'skills'), `pi-package:${path.basename(packageRoot)}`);
    }

    const seen = new Set();
    return candidates.filter(({ root }) => {
      let realRoot;
      try {
        realRoot = fs.realpathSync(root);
      } catch {
        return false;
      }
      if (seen.has(realRoot)) return false;
      seen.add(realRoot);
      return true;
    });
  }

  refresh() {
    const skills = new Map();

    for (const { root, source } of this.candidateRoots()) {
      let realRoot;
      try {
        realRoot = fs.realpathSync(root);
      } catch {
        continue;
      }

      for (const entry of fs.readdirSync(root, { withFileTypes: true })) {
        if (!entry.isDirectory()) continue;
        const skillDir = path.join(root, entry.name);
        const skillFile = path.join(skillDir, 'SKILL.md');
        if (!fs.existsSync(skillFile)) continue;

        try {
          const realDir = fs.realpathSync(skillDir);
          const realSkillFile = fs.realpathSync(skillFile);
          if (!isWithin(realRoot, realDir) || !isWithin(realDir, realSkillFile)) continue;

          const content = fs.readFileSync(realSkillFile, 'utf8');
          const metadata = parseFrontmatter(content, entry.name);
          if (!metadata.name || skills.has(metadata.name)) continue;

          skills.set(metadata.name, {
            name: metadata.name,
            description: metadata.description,
            dir: realDir,
            file: realSkillFile,
            source
          });
        } catch {
          // Ignore unreadable or invalid skill directories and continue discovery.
        }
      }
    }

    this.skills = skills;
    return this.list(false);
  }

  list(refresh = true) {
    if (refresh) this.refresh();
    return Array.from(this.skills.values())
      .sort((a, b) => a.name.localeCompare(b.name))
      .map(skill => ({ ...skill }));
  }

  catalogText({ refresh = true, maxDescription = 220 } = {}) {
    const skills = this.list(refresh);
    if (!skills.length) return '(No local skills discovered.)';
    return skills.map(skill => {
      const description = skill.description
        ? skill.description.slice(0, maxDescription) + (skill.description.length > maxDescription ? '…' : '')
        : 'No description provided.';
      return `- ${skill.name}: ${description}`;
    }).join('\n');
  }

  read(skillName, relativePath = 'SKILL.md') {
    this.refresh();
    const skill = this.skills.get(String(skillName || '').trim());
    if (!skill) {
      throw new Error(`Unknown skill '${skillName}'. Call list_skills to inspect available skills.`);
    }

    const requestedPath = String(relativePath || 'SKILL.md').trim() || 'SKILL.md';
    if (path.isAbsolute(requestedPath)) {
      throw new Error('Skill file path must be relative to the skill directory.');
    }

    const target = path.resolve(skill.dir, requestedPath);
    if (!isWithin(skill.dir, target)) {
      throw new Error('Access denied: Skill path escapes the skill directory.');
    }
    if (!fs.existsSync(target)) {
      throw new Error(`Skill file not found: ${requestedPath}`);
    }

    const realTarget = fs.realpathSync(target);
    if (!isWithin(skill.dir, realTarget)) {
      throw new Error('Access denied: Skill file resolves outside the skill directory.');
    }

    const stat = fs.statSync(realTarget);
    if (!stat.isFile()) {
      throw new Error(`Skill path is not a file: ${requestedPath}`);
    }
    if (stat.size > MAX_SKILL_FILE_BYTES) {
      throw new Error(`Skill file is too large (${stat.size} bytes; max ${MAX_SKILL_FILE_BYTES}).`);
    }

    return {
      skill,
      requestedPath,
      absolutePath: realTarget,
      bytes: stat.size,
      content: fs.readFileSync(realTarget, 'utf8')
    };
  }
}

module.exports = SkillRegistry;

File: src-tauri/src/skills.rs
use serde::Serialize;
use serde_yaml::Value as YamlValue;
use std::{
    collections::{HashMap, HashSet},
    fs,
    path::{Path, PathBuf},
};

const MAX_SKILL_FILE_BYTES: u64 = 1024 * 1024;

#[derive(Debug, Clone, Serialize)]
pub struct SkillInfo {
    pub name: String,
    pub description: String,
    pub dir: PathBuf,
    pub source: String,
}

fn home_dir() -> PathBuf {
    dirs::home_dir().unwrap_or_else(|| PathBuf::from("."))
}

fn expand_home(value: &str) -> PathBuf {
    let trimmed = value.trim();
    if trimmed.is_empty() || trimmed == "~" {
        return home_dir();
    }
    if let Some(rest) = trimmed.strip_prefix("~/") {
        return home_dir().join(rest);
    }
    PathBuf::from(trimmed)
}

fn direct_package_roots(node_modules: &Path) -> Vec<PathBuf> {
    let mut roots = Vec::new();
    let Ok(entries) = fs::read_dir(node_modules) else {
        return roots;
    };
    for entry in entries.flatten() {
        let Ok(file_type) = entry.file_type() else {
            continue;
        };
        if !file_type.is_dir() {
            continue;
        }
        let name = entry.file_name().to_string_lossy().to_string();
        if name.starts_with('.') {
            continue;
        }
        if name.starts_with('@') {
            if let Ok(scoped) = fs::read_dir(entry.path()) {
                for package in scoped.flatten() {
                    if package
                        .file_type()
                        .map(|kind| kind.is_dir())
                        .unwrap_or(false)
                    {
                        roots.push(package.path());
                    }
                }
            }
        } else {
            roots.push(entry.path());
        }
    }
    roots
}

fn parse_frontmatter(content: &str, fallback: &str) -> (String, String) {
    let normalized = content.replace("\r\n", "\n");
    let Some(rest) = normalized.strip_prefix("---\n") else {
        return (fallback.to_string(), String::new());
    };
    let Some(end) = rest.find("\n---") else {
        return (fallback.to_string(), String::new());
    };
    let metadata: YamlValue = serde_yaml::from_str(&rest[..end]).unwrap_or(YamlValue::Null);
    let name = metadata
        .get("name")
        .and_then(YamlValue::as_str)
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .unwrap_or(fallback)
        .to_string();
    let description = metadata
        .get("description")
        .and_then(YamlValue::as_str)
        .unwrap_or_default()
        .split_whitespace()
        .collect::<Vec<_>>()
        .join(" ");
    (name, description)
}

fn candidate_roots(fs_root: &str) -> Vec<(PathBuf, String)> {
    let home = home_dir();
    let mut roots = vec![
        (
            expand_home(fs_root).join(".agents/skills"),
            "workspace".to_string(),
        ),
        (home.join(".agents/skills"), "user".to_string()),
        (home.join(".pi/agent/skills"), "pi".to_string()),
    ];

    let extensions = home.join(".pi/agent/extensions");
    if let Ok(entries) = fs::read_dir(extensions) {
        for entry in entries.flatten() {
            if entry.file_type().map(|kind| kind.is_dir()).unwrap_or(false) {
                let name = entry.file_name().to_string_lossy().to_string();
                roots.push((entry.path().join("skills"), format!("pi-extension:{name}")));
            }
        }
    }

    for package in direct_package_roots(&home.join(".pi/agent/npm/node_modules")) {
        let name = package
            .file_name()
            .map(|value| value.to_string_lossy().to_string())
            .unwrap_or_else(|| "package".into());
        roots.push((package.join("skills"), format!("pi-package:{name}")));
    }

    roots
}

pub fn discover(fs_root: &str) -> Vec<SkillInfo> {
    let mut skills = HashMap::<String, SkillInfo>::new();
    let mut seen_roots = HashSet::new();

    for (root, source) in candidate_roots(fs_root) {
        let Ok(real_root) = fs::canonicalize(root) else {
            continue;
        };
        if !seen_roots.insert(real_root.clone()) {
            continue;
        }
        let Ok(entries) = fs::read_dir(&real_root) else {
            continue;
        };
        for entry in entries.flatten() {
            if !entry.file_type().map(|kind| kind.is_dir()).unwrap_or(false) {
                continue;
            }
            let skill_dir = entry.path();
            let skill_file = skill_dir.join("SKILL.md");
            let (Ok(real_dir), Ok(real_file)) =
                (fs::canonicalize(&skill_dir), fs::canonicalize(&skill_file))
            else {
                continue;
            };
            if !real_dir.starts_with(&real_root) || !real_file.starts_with(&real_dir) {
                continue;
            }
            let Ok(content) = fs::read_to_string(&real_file) else {
                continue;
            };
            let fallback = entry.file_name().to_string_lossy().to_string();
            let (name, description) = parse_frontmatter(&content, &fallback);
            if name.is_empty() || skills.contains_key(&name) {
                continue;
            }
            skills.insert(
                name.clone(),
                SkillInfo {
                    name,
                    description,
                    dir: real_dir,
                    source: source.clone(),
                },
            );
        }
    }

    let mut result = skills.into_values().collect::<Vec<_>>();
    result.sort_by(|a, b| a.name.cmp(&b.name));
    result
}

pub fn read(fs_root: &str, skill_name: &str, relative_path: &str) -> Result<String, String> {
    let skills = discover(fs_root);
    let skill = skills
        .into_iter()
        .find(|skill| skill.name == skill_name.trim())
        .ok_or_else(|| format!("Unknown skill '{}'. Call list_skills first.", skill_name))?;

    let requested = if relative_path.trim().is_empty() {
        "SKILL.md"
    } else {
        relative_path.trim()
    };
    let requested_path = Path::new(requested);
    if requested_path.is_absolute() {
        return Err("Skill file path must be relative to the skill directory.".into());
    }

    let target = path_clean::PathClean::clean(&skill.dir.join(requested_path));
    if !target.starts_with(&skill.dir) {
        return Err("Access denied: Skill path escapes the skill directory.".into());
    }
    let real_target =
        fs::canonicalize(&target).map_err(|_| format!("Skill file not found: {requested}"))?;
    if !real_target.starts_with(&skill.dir) {
        return Err("Access denied: Skill file resolves outside the skill directory.".into());
    }
    let metadata = fs::metadata(&real_target).map_err(|error| error.to_string())?;
    if !metadata.is_file() {
        return Err(format!("Skill path is not a file: {requested}"));
    }
    if metadata.len() > MAX_SKILL_FILE_BYTES {
        return Err(format!(
            "Skill file is too large ({} bytes; max {}).",
            metadata.len(),
            MAX_SKILL_FILE_BYTES
        ));
    }
    fs::read_to_string(real_target).map_err(|error| error.to_string())
}
